core/src/main/java/fj/data/Stream.java

```diff
@@ -386,13 +386,7 @@ public final void foreachDoEffect(final Effect1<A> f) {
    * @return A new stream after performing the map, then final join.
    */
   public final <B> Stream<B> bind(final F<A, Stream<B>> f) {
-    return map(f).foldLeft((accumulator, element) -> {
-        Stream<B> result = accumulator;
-        for (B single : element) {
-            result = result.cons(single);
-        }
-        return result;
-    }, Stream.<B>nil()).reverse();
+    return foldRight(h -> (t -> f.f(h).append(t)), nil());
   }
```

Comment:
Indeed, foldRight uses O(n) space in worst case. But in this case it doesn't, because Stream.append() is lazy!
If you are not convinced, here is an example, which fails with stackoverflow with the previous implementation and doesn't fail with the new implementation:
```java
for (int j : Stream.forever(Enumerator.intEnumerator, 0).bind(i -> Stream.stream(i, i)).take(10000)) {
  System.out.println(j); // prints 0, 0, 1, 1, 2, 2, 3, 3, ...
}
```
If you are still not convinced, scalaz uses the same implementation of bind: https://github.com/scalaz/scalaz/blob/v7.2.0-M1/core/src/main/scala/scalaz/EphemeralStream.scala#L64